Give ThriftTransportPool higher concurrent granularity

lbschanno opened the pull request:
~~Note: Draft PR only, please do not merge.~~

This is my initial approach to fixing #1122. I'm making this available here for review and suggestions. I am in the process of creating unit tests as well.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -67,7 +70,8 @@ public CachedConnection reserveAny() {
-  private Map<ThriftTransportKey,CachedConnections> cache = new HashMap<>();
+  private ConcurrentHashMap<ThriftTransportKey,Lock> locks = new ConcurrentHashMap<>();
+  private ConcurrentHashMap<ThriftTransportKey,CachedConnections> cache = new ConcurrentHashMap<>();
   private Map<ThriftTransportKey,Long> errorCount = new HashMap<>();
```
These error tracking maps and sets are accessed while different locks are held.  I created lbschanno/accumulo#2 as one possible way to bring those under the same lock.

lbschanno replied:
Okay, looks good to me. I'll merge the changes in.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -564,9 +556,12 @@ public void returnTransport(TTransport tsc) {
     }
 
     // close outside of sync block
+    // TODO - Might need to close within lock block.
```
Understood.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -649,14 +644,65 @@ public synchronized void startCheckerThread() {
+      Lock lock = getLock(entry.getKey());
+      lock.lock();
+
+      CachedConnections cachedConns = entry.getValue();
```
Agreed, good catch.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -413,14 +382,20 @@ public TTransport getTransport(HostAndPort location, long milliseconds, ClientCo
       CachedConnection cachedConnection =
           getCache().computeIfAbsent(cacheKey, ck -> new CachedConnections()).reserveAny();
       if (cachedConnection != null) {
         log.trace("Using existing connection to {}", cacheKey.getServer());
```
Okay. In the meantime we can move this check and logging outside of the try-finally block.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -67,7 +70,8 @@ public CachedConnection reserveAny() {
   }
 
-  private Map<ThriftTransportKey,CachedConnections> cache = new HashMap<>();
+  private ConcurrentHashMap<ThriftTransportKey,Lock> locks = new ConcurrentHashMap<>();
```
I think these need to be volatile. Also a single volatile pointer to `locks` and `cache` may help with the fact that shutdown can now run concurrently with connection reservation and allocation. 

Maybe something like 

```java
   private static class LocksAndCache {
      ConcurrentHashMap<ThriftTransportKey,Lock> locks = new ConcurrentHashMap<>();
      ConcurrentHashMap<ThriftTransportKey,CachedConnections> cache = new ConcurrentHashMap<>();
   }

  private volatile LocksAndCache locksAndCache = new LocksAndCahce();
```

I'll explain in other comments why I think this may be useful.

lbschanno replied:
I created something similar to this, but called `ConnectionPool`. I also added on additional methods both for `ConnectionPool` and `CachedConnections` in an attempt to make the code more concise overall.

keith-turner replied:
I like the changes you made.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -690,4 +753,8 @@ private void shutdown() {
     return cache;
   }
+
+  private Lock getLock(final ThriftTransportKey key) {
```
I think the getLock() and getCache() methods could be replaced with a single method like the following

```java
  private LocksAndCache getLocksAndCache() {
    // very important to only read volatile once into a local variable as multiple reads may see different things.
    LocksAndCache local  = locksAndCache;
    if (local == null)
      throw new TransportPoolShutdownException(
          "The Accumulo singleton for connection pooling is disabled.  This is likely caused by "
              + "all AccumuloClients being closed or garbage collected.");
    return local;
  }
```
The code should be analyzed to ensure external operations only retrieve this once. Or if it is retrieved multiple times, analysis should be done to ensure the code handles changes caused by shutdown.  

Any code retrieving this should properly handle a concurrent shutdown.  Some code already does this.

lbschanno replied:
Made this changed, but the method is `getConnectionPool()` instead. I believe all calls to this are handling concurrent shutdown, with the exception of `returnTransport()` which I'm still not entirely sure about.

keith-turner replied:
I need to look at returnTransport() in the changes I made.

keith-turner replied:
`returnTransport()` LGTM.  I did make a comment elsewhere about moving access to the map outside of the lock.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -498,10 +485,15 @@ private TTransport createNewTransport(ThriftTransportKey cacheKey) throws TTrans
+        lock.unlock();
       }
     } catch (TransportPoolShutdownException e) {
       cc.transport.close();
```
This is an example of code that is handles a concurrent shutdown.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -520,27 +512,17 @@ public void returnTransport(TTransport tsc) {
     ArrayList<CachedConnection> closeList = new ArrayList<>();
 
-    synchronized (this) {
+    Lock lock = getLock(ctsc.getCacheKey());
```
This is at least once place that probably needs to handle concurrent shutdown.  If we try to get the locks and cache and its shutdown, then maybe jsut close the connection?

lbschanno replied:
I'm not entirely certain that I managed to address this sufficiently, but let me know what you think.

keith-turner replied:
May have handled this with the changes in lbschanno#3

keith-turner commented:
@lbschanno I really like the refactoring you did in 20a9e63.  I opened lbschanno/accumulo#3 as an attempt to further simplify the locking and synchronization.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,9 +68,167 @@ public CachedConnection reserveAny() {
+      return connections.computeIfAbsent(key, k -> new CachedConnections());
+    }
+
+    boolean unreserveConnection(final CachedTTransport transport,
```
When reviewing the code, I found myself wondering if some of these functions were called with a lock?  It made me wonder if these functions that are only called using `executeWithinLock` should be inlined.  However inlining may look awful.  I am not sure about this, just sharing my experience.

lbschanno replied:
My primary intent was to ensure that any action that needed to execute within a lock would automatically do so. The lack of inlining was a side benefit. However, if you feel it would improve the overall clarity to inline instead, I can switch it. Alternatively, I can add some documentation for the methods in question to explicitly state that their operations take place within a lock.

keith-turner replied:
> Alternatively, I can add some documentation for the methods in question to explicitly state that their operations take place within a lock.

I like that better than inlining.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,9 +68,167 @@ public CachedConnection reserveAny() {
+
+    <T> T executeWithinLock(final ThriftTransportKey key, Supplier<T> function) {
+      Lock lock = getLock(key);
+      try {
```
Handled this lbschanno#3

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,9 +68,167 @@ public CachedConnection reserveAny() {
+    }
+
+    void shutdown() {
+      for (Entry<ThriftTransportKey,CachedConnections> entry : connections.entrySet()) {
```
I think there are race conditions where this loop will not see all of the locks (like locks added during the lifetime of the loop).  That is why I used an array of locks in lbschanno#3

lbschanno replied:
Agreed, I merged in your changes with the array of locks.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,14 +69,206 @@ public CachedConnection reserveAny() {
+
+    CachedConnection reserveAnyIfPresent(final ThriftTransportKey key) {
+      return executeWithinLock(key,
+          () -> connections.containsKey(key) ? connections.get(key).reserveAny() : null);
```
Sounds good, I'll add the change.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,14 +69,206 @@ public CachedConnection reserveAny() {
+    }
+
+    CachedConnection reserveAny(final ThriftTransportKey key) {
+      return executeWithinLock(key, () -> getOrCreateCachedConnections(key).reserveAny());
```
Yes, currently keys are never removed from the concurrent hash map, so we don't need to worry about a conflict there. I think your recommendation should work safely.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,14 +69,206 @@ public CachedConnection reserveAny() {
+
+    void putReserved(final ThriftTransportKey key, final CachedConnection connection) {
+      executeWithinLock(key,
+          () -> getOrCreateCachedConnections(key).reserved.put(connection.transport, connection));
```
Agreed.

lbschanno commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -65,14 +69,206 @@ public CachedConnection reserveAny() {
+
+    boolean unreserveConnection(final CachedTTransport transport,
+        final List<CachedConnection> toBeClosed) {
+      CachedConnections connections = getCachedConnections(transport.getCacheKey());
```
Agreed.

lbschanno commented:
I added the additional suggestions, and then rebased in order to update the branch with all commits currently in master, and to consolidate all of the PR-related commits into a single commit.

@keith-turner I can to write an IT test to test the overall correctness of the functions that we changed, but I'm not entirely certain how to write tests to ensure the correctness of the multithreading aspects.

lbschanno commented:
Please hold of on merging this in, I forgot to do a proper build to ensure the files are formatted correctly. Currently doing that now.

keith-turner commented:
@lbschanno testing that single threaded functionality is correct would be nice.  Ideally mocking could be used in a unit test, so an IT is not needed.  Do you want to do that in a follow up PR?

As for multi-threaded test, I think a simple test could be written.  Like have 10 threads all request connections and make sure they all get diff connections.  Could use a barrier (count down latch), where each thread does something like the following.

```java
   var conn  = getConnection(...)
   sharedList1.add(conn)
   latch1.countDown()
   latch1.await()
   returnConn(conn);

   conn  = getConnection(...)
   sharedList2.add(conn)
   latch2.countDown()
   latch2.await()
   returnConn(conn);
```

in the main test method could do the following

```java
  // create the two countdown latches of size X
  // create the two shared list
  // create X threads each running the code above
  latch1.await()
  // verify there are X unique objects in sharedList1
  latch2.await()
  // verify there are X unique objects in sharedList2 
```

lbschanno commented:
Forced pushed an edit to the commit that includes formatting changes that resulted from a build.

lbschanno commented:
@keith-turner I think we can do the tests in a follow-up PR. Thanks for the suggestions on structuring the multi-threaded tests, I feel I can handle doing that.

lbschanno commented:
Force-pushed an edit to the commits to remove trailing whitespace and to address the following errors from FindBugs.
```
 [ERROR] Bad attempt to compute absolute value of signed 32-bit hashcode in org.apache.accumulo.core.clientImpl.ThriftTransportPool$ConnectionPool.getLock(ThriftTransportKey) [org.apache.accumulo.core.clientImpl.ThriftTransportPool$ConnectionPool] At ThriftTransportPool.java:[line 265] RV_ABSOLUTE_VALUE_OF_HASHCODE
 [ERROR] org.apache.accumulo.core.clientImpl.ThriftTransportPool$ConnectionPool.shutdown() does not release lock on all paths [org.apache.accumulo.core.clientImpl.ThriftTransportPool$ConnectionPool] At ThriftTransportPool.java:[line 228] UL_UNRELEASED_LOCK
```
 I modified the absolute value computation, and suppressed the UL_UNRELEASED_LOCK error.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java:
```diff
@@ -44,8 +50,11 @@
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@SuppressFBWarnings(value = "UL_UNRELEASED_LOCK",
```
Do you know if moving this suppression to the shutdown() method would work?  If so, making it more narrow would be nice.

lbschanno replied:
Yes it does apparently. Just moved it.

ctubbsii commented:
@lbschanno It's generally best to avoid force-pushing, until the end. When we merge, we usually squash the whole PR into one commit (unless there's a good reason not to), so having multiple smaller commits, rather than amending and force-pushing, isn't a problem. Actually, it's easier to see what has changed since the last time looking at the PR, when there are multiple commits, rather than force-push.

lbschanno commented:
@ctubbsii ah I see, I didn't know the PR is usually squashed when merged in. I'll keep that in mind.
